contracts/crowdsale/crowdsale-mintable.aes

```diff
@@ -0,0 +1,165 @@
+
+        _updatePurchasingState(_beneficiary, _amount)
+
+        _forwardFunds(_amount)
+        _postValidatePurchase(_beneficiary, _amount)
+
+    public function endCrowdsale() =
+        isOwner()
+        state.erc20token.transferOwnership(Call.caller)
+
+    public function getEndDate() : int =
+        state.endDate
+
+    // Validation of an incoming purchase. Use require statements to revert state when conditions are not met.
+    // _beneficiary Address performing the token purchase
+    // _amount Value involved in the purchase
+    private function _preValidatePurchase(_beneficiary : address , _amount : int) =
+            require(_beneficiary != #0, "Should not be empty address.")
+            require(_amount != 0, "Value cannot be zero.")
+
+    // Validation of an executed purchase. Observe state and use revert statements to undo rollback when valid conditions are not met.
+    // _beneficiary Address performing the token purchase
+    // _amount Value involved in the purchase
+    private function _postValidatePurchase(_beneficiary : address, _amount : int) : bool =
+        require(true, "Error message")
+
+    // Source of tokens. Override this method to modify the way in which the crowdsale ultimately gets and sends its tokens.
+    // _beneficiary Address performing the token purchase
+    // _tokenAmount Number of tokens to be emitted
+    private function _deliverTokens(_beneficiary : address, _tokenAmount : int, _token : ERC20) =
+        _token.mint(_beneficiary, _tokenAmount)
+    
+    // Executed when a purchase has been validated and is ready to be executed. Doesn't necessarily emit/send tokens.
+    // _beneficiary Address receiving the tokens
+    // _tokenAmount Number of tokens to be purchased
+    private function _processPurchase(_beneficiary : address, _tokenAmount : int) =
+        _deliverTokens(_beneficiary, _tokenAmount, state.erc20token)
+
+    // Override for extensions that require an internal state to check for validity (current user contributions, etc.)
+    // _beneficiary Address receiving the tokens
+    // _amount Value in wei involved in the purchase
+    private function _updatePurchasingState(_beneficiary : address, _amount : int) =
+        // optional override
+        require(true, "Error message")
+
+
+    // Override to extend the way in which ether is converted to tokens.
+    // _amount Value to be converted into tokens
+    //  Number of tokens that can be purchased with the specified _weiAmount
+    private function _getTokenAmount(_amount : int, _rate : int) : int =
+        mul(_amount, _rate)
+
+    private function _forwardFunds(_amount : int) =
+        Chain.spend(state.wallet, _amount)
+
+
+    private function require(expression : bool, error : string) =
+        if(!expression) 
+            abort(error)
+
+        true
+
+    private function isCrowdsale() = 
+        if(Chain.timestamp > state.endDate)
```

Comment:
I think this method can be substituted via `require(Chain.timestamp <= state.endDate, "Crowdsale has ended!")`